scheme/src/main/java/cz/o2/proxima/scheme/avro/AvroSerializerFactory.java

```diff
@@ -0,0 +1,125 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.scheme.avro;
+
+
+import cz.o2.proxima.scheme.SerializationException;
+import cz.o2.proxima.scheme.ValueSerializer;
+import cz.o2.proxima.scheme.ValueSerializerFactory;
+import cz.o2.proxima.util.Classpath;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.avro.Schema;
+import org.apache.avro.specific.SpecificRecord;
+
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.URI;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+
+/**
+ * Avro serializer factory for manipulate with SpecificRecords
+ */
+@Slf4j
+public class AvroSerializerFactory implements ValueSerializerFactory {
+  private final Map<URI, ValueSerializer<?>> serializersCache = new ConcurrentHashMap<>();
+
+  @Override
+  public String getAcceptableScheme() {
+    return "avro";
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T> ValueSerializer<T> getValueSerializer(URI specifier) {
+    return (ValueSerializer<T>) serializersCache.computeIfAbsent(
+        specifier, this::createSerializer);
+  }
+
+  private <M extends SpecificRecord> ValueSerializer<M> createSerializer(URI uri) {
+    return new ValueSerializer<M>() {
+      final String avroClassName = uri.getSchemeSpecificPart();
+      @SuppressWarnings("unchecked")
+      final Class<M> clz = (Class<M>) Classpath
+          .findClass(avroClassName, SpecificRecord.class);
+
+      transient AvroSerializer<M> avroSerializer = null;
+
+      @Override
+      public Optional<M> deserialize(byte[] input) {
+        if (input.length == 0) {
+          /**
+           * This is a little bit weird but proxima try to deserialize empty bytes
+           * in serializer validation {@link ValueSerializer#isValid(byte[])}
+           */
+          return Optional.of(getDefault());
+        }
+        if (avroSerializer == null) {
+          avroSerializer = new AvroSerializer<>(getAvroSchemaForClass(avroClassName));
+        }
+        try {
+          return Optional.of(avroSerializer.deserialize(input));
+        } catch (IOException ex) {
+          throw new SerializationException("Unable to deserialize avro payload", ex);
+        }
+      }
+
+      @Override
+      public byte[] serialize(M value) {
+        if (avroSerializer == null) {
+          avroSerializer = new AvroSerializer<>(getAvroSchemaForClass(avroClassName));
+        }
+        try {
+          return avroSerializer.serialize(value);
+        } catch (IOException ex) {
+          throw new SerializationException("Unable to serialize avro object", ex);
+        }
+      }
+
+      @Override
+      public M getDefault() {
+        try {
+          return clz.newInstance();
```

Comment:
You can use `Classpath.newInstance()`, which will catch and retrow exceptions for you.

Reply:
And also - it will be better to create this in constructor.

Reply:
Fixed